pkg/js/validator/validate-store.ts

```diff
@@ -0,0 +1,976 @@
+      for (const assertion of Object.keys(listUsers.assertions)) {
+        if (listUsers.assertions[assertion].users) {
+          for (const user of listUsers.assertions[assertion].users) {
+            if (!tuples.some((tuple) => tuple.user === user)) {
+              errors.push(undefinedTypeTuple(user, instancePath + `/assertions/${assertion}/users`));
+            }
+          }
+        }
+      }
+    }
+  }
+
+  // Check context params
+  if (listUsers.context) {
+    for (const testParam in listUsers.context) {
+      if (!params.includes(testParam)) {
+        errors.push(unidentifiedTestParam(testParam, instancePath + `/context/${testParam}`));
+      }
+    }
+  }
+
+  return errors;
+}
+
+// Validation for types in check
+function validateTestTypes(store: Store, model: AuthorizationModel, instancePath: string): boolean {
+  const errors = [];
+
+  // Collect params for validity checking
+  const params: string[] = [];
+  for (const condition in model.conditions) {
+    for (const param in model.conditions[condition].parameters) {
+      params.push(param);
+    }
+  }
+
+  for (const testNo in store.tests) {
+    // Collect valid tuples
+    const tuples = [];
+    if (store.tuples && store.tuples.length) {
+      tuples.push(...store.tuples);
+    }
+
+    const test = store.tests[testNo];
+    if (!test) {
+      continue;
+    }
+
+    if (test.tuples && test.tuples.length) {
+      tuples.push(...test.tuples);
+    }
+
+    // Validate check
+    for (const checkNo in test.check) {
+      if (!test.check[checkNo] || !test.check[checkNo].user || !test.check[checkNo].object) {
+        return false;
+      }
+      errors.push(
+        ...validateCheck(
+          model,
+          test.check[checkNo],
+          tuples,
+          params,
+          instancePath + `/tests/${testNo}/check/${checkNo}`,
+        ),
+      );
+    }
+
+    // Validate list objects
+    for (const listNo in test.list_objects) {
+      if (!test.list_objects[listNo] || !test.list_objects[listNo].user || !test.list_objects[listNo].type) {
+        return false;
+      }
+      errors.push(
+        ...validateListObject(
+          model,
+          test.list_objects[listNo],
+          tuples,
+          params,
+          instancePath + `/tests/${testNo}/list_objects/${listNo}`,
+        ),
+      );
+    }
+
+    for (const listNo in test.list_users) {
+      if (!test.list_users[listNo] || !test.list_users[listNo].object || !test.list_users[listNo].user_filter) {
+        return false;
+      }
+      errors.push(
+        ...validateListUsers(
+          model,
+          test.list_users[listNo],
+          tuples,
+          params,
+          instancePath + `/tests/${testNo}/list_users/${listNo}`,
+        ),
+      );
+    }
+  }
+
+  if (errors.length) {
+    validateStore.errors?.push(...errors);
+    return false;
+  }
+  return true;
+}
+
+const validateStore: SchemaValidateFunction = function (
+  this: { jsonModel: AuthorizationModel },
```

Comment:
this and ctx, can we have better names?

Reply:
`jsonModel` is needed for both the validation of the overall store, as well as the tuples.

I'm setting `this` here, via `call`, such that the `jsonModel` is persisted for all validation being referenced by `validateStore` & `validateTuples`. 

Doing so means we dont have to rebuild the schema validator every time we want to invoke  the validation.

I've tried today to do this several alternatives today, and I cant find a way to provide the model data.